railib/api.py

```diff
@@ -657,43 +649,83 @@ def _model(name: str, model: str) -> dict:
     }
 
 
-# Returns full list of models.
-def _list_models(ctx: Context, database: str, engine: str) -> dict:
-    tx = Transaction(database, engine, mode=Mode.OPEN)
-    rsp = tx.run(ctx, _list_action())
-    actions = rsp["actions"]
-    assert len(actions) == 1
-    action = actions[0]
-    models = action["result"]["sources"]
-    return models
-
-
 def create_database(ctx: Context, database: str, source: str = None) -> dict:
     data = {"name": database, "source_name": source}
     url = _mkurl(ctx, PATH_DATABASE)
     rsp = rest.put(ctx, url, data)
     return json.loads(rsp.read())
 
 
-def delete_model(ctx: Context, database: str, engine: str, model: str) -> dict:
-    tx = Transaction(database, engine, mode=Mode.OPEN, readonly=False)
-    actions = [_delete_model_action(model)]
-    return tx.run(ctx, *actions)
+# Returns full list of models.
+def list_models(ctx: Context, database: str, engine: str) -> List:
+    models = []
+    out_name = f'model{random.randint(0, sys.maxsize)}'
+    resp = exec(ctx, database, engine, f'def output:{out_name}[name] = rel:catalog:model(name, _)')
+    for result in resp.results:
+        if f'/:output/:{out_name}' in result['relationId']:
+            table = result['table'].to_pydict()
+            models.extend([table['v1'][i] for i in range(1, len(table['v1']))])
+
+    return models
+
+
+def delete_models(ctx: Context, database: str, engine: str, models: List[str]) -> TransactionAsyncResponse:
+    queries = [
+        f'def delete:rel:catalog:model["{model_name}"] = rel:catalog:model["{model_name}"]'
+        for model_name in models
+    ]
+    return exec(ctx, database, engine, '\n'.join(queries), readonly=False)
+
+
+def delete_models_async(ctx: Context, database: str, engine: str, models: List[str]) -> TransactionAsyncResponse:
+    queries = [
+        f'def delete:rel:catalog:model["{model_name}"] = rel:catalog:model["{model_name}"]'
+        for model_name in models
+    ]
+    return exec_async(ctx, database, engine, '\n'.join(queries), readonly=False)
 
 
 # Returns the named model
 def get_model(ctx: Context, database: str, engine: str, name: str) -> str:
-    models = _list_models(ctx, database, engine)
-    for model in models:
-        if model["name"] == name:
-            return model["value"]
+    out_name = f'model{random.randint(0, sys.maxsize)}'
+    cmd = f'def output:{out_name} = rel:catalog:model["{name}"]'
+    resp = exec(ctx, database, engine, cmd)
+    for result in resp.results:
+        if f'/:output/:{out_name}' in result['relationId']:
```

Comment:
ditto

Reply:
that's the reason why we generated a random output name then we filter out results based on that
users can install predefined output relations that could collapse with the `get_model` one